staging/src/k8s.io/component-helpers/scheduling/corev1/helpers.go

```diff
@@ -30,3 +35,109 @@ func PodPriority(pod *v1.Pod) int32 {
 	// name of the pod was empty. So, we resolve to the static default priority.
 	return 0
 }
+
+// NodeSelectorRequirementsAsSelector converts the []NodeSelectorRequirement API type into a labels.Selector.
+func nodeSelectorRequirementsAsSelector(nsr []v1.NodeSelectorRequirement) (labels.Selector, error) {
+	if len(nsr) == 0 {
+		return labels.Nothing(), nil
+	}
+	selector := labels.NewSelector()
+	for _, expr := range nsr {
+		var op selection.Operator
+		switch expr.Operator {
+		case v1.NodeSelectorOpIn:
+			op = selection.In
+		case v1.NodeSelectorOpNotIn:
+			op = selection.NotIn
+		case v1.NodeSelectorOpExists:
+			op = selection.Exists
+		case v1.NodeSelectorOpDoesNotExist:
+			op = selection.DoesNotExist
+		case v1.NodeSelectorOpGt:
+			op = selection.GreaterThan
+		case v1.NodeSelectorOpLt:
+			op = selection.LessThan
+		default:
+			return nil, fmt.Errorf("%q is not a valid node selector operator", expr.Operator)
+		}
+		r, err := labels.NewRequirement(expr.Key, op, expr.Values)
+		if err != nil {
+			return nil, err
+		}
+		selector = selector.Add(*r)
+	}
+	return selector, nil
+}
+
+// MatchNodeSelectorTerms checks whether the node labels or name match node selector terms in ORed;
+// nil or empty term matches no objects.
+func MatchNodeSelectorTerms(
+	node *v1.Node,
+	nodeSelectorTerms []v1.NodeSelectorTerm,
+) (bool, error) {
+	errors := make([]error, 0)
+	for _, req := range nodeSelectorTerms {
+		// nil or empty term selects no objects
+		if len(req.MatchExpressions) == 0 && len(req.MatchFields) == 0 {
+			continue
+		}
+
+		if len(req.MatchExpressions) != 0 {
+			labelSelector, err := nodeSelectorRequirementsAsSelector(req.MatchExpressions)
```

Comment:
I didn't notice this, but nothing was reusing these processed selectors, right?

We might want to do that for the node affinity plugin:

```
TODO: Avoid computing it for all nodes if this becomes a performance problem.
```

If we are not introducing any regression at this point, we could leave that for later. Then we can add:

```
MatchParsedNodeSelectorTerms(*v1.Node, []ParsedNodeSelectorTerm) bool
```

Reply:
Since this is intended to be used by more than just the scheduler, wouldn't it make sense to keep the processing in this function itself? Otherwise 3rd party callers will have to implement `nodeSelectorRequirementsAsFieldSelector` and `nodeSelectorRequirementsAsSelector` themselves (or we make those functions exportable). Though I agree this could be fixed later perhaps calculated in preScore/preFilter

Reply:
we can think of that later.

Reply:
> wouldn't it make sense to keep the processing in this function itself

The point of the TODO was that if you have a selector you are going to evaluate against many nodes, constructing the selector per-node is expensive. If the function is limited to a single node, you cannot do the shared selector computation inside that function.

Reply:
In the kube-scheduler's use case that makes sense, where we evaluate every node as a basic requirement of scheduling. But if the point of externalizing this helper is to make it accessible by 3rd party components and simplify the function signature, requiring preprocessing of the input selectors feels like it defeats the goal. We can do that, but like I said above we should also export the `requirementsAsSelector` helpers in that case.